File: backend/app/schemas/load_demand.py
"""
Pydantic schemas for electrical load and demand calculations.
KESE-S2-M2
"""

from decimal import Decimal
from typing import Annotated, Self

from pydantic import (
    BaseModel,
    BeforeValidator,
    ConfigDict,
    Field,
    StrictInt,
    model_validator,
)

from app.domain.electrical.jurisdiction.jurisdiction_models import (
    JurisdictionProfile,
)
from app.domain.electrical.loads.models import (
    LoadGroupInput,
    LoadInput,
    LoadScenario,
    PhaseSystem,
    PowerBasis,
)
from app.domain.electrical.loads.results import (
    CalculationStatus,
    LoadGroupCalculationResult,
    LoadWarningCode,
)


def _reject_float(value: object) -> object:
    """Reject binary floating-point engineering inputs."""

    if isinstance(value, float):
        raise ValueError(
            "engineering decimal values must be provided as strings, integers, or Decimal values"
        )

    return value


type ExactDecimal = Annotated[
    Decimal,
    BeforeValidator(_reject_float),
]


class _RequestBase(BaseModel):
    """Shared API request configuration."""

    model_config = ConfigDict(
        extra="forbid",
        str_strip_whitespace=True,
    )


class _ResponseBase(BaseModel):
    """Shared API response configuration."""

    model_config = ConfigDict(
        from_attributes=True,
        extra="forbid",
    )


class LoadCalculationRequest(_RequestBase):
    """API request for calculating one electrical load."""

    code: str = Field(
        min_length=1,
        max_length=50,
    )
    name: str = Field(
        min_length=1,
        max_length=200,
    )
    quantity: StrictInt = Field(
        gt=0,
    )
    rated_power_kw: ExactDecimal = Field(
        gt=Decimal("0"),
        max_digits=38,
        decimal_places=18,
    )
    phase_system: PhaseSystem
    voltage_v: ExactDecimal = Field(
        gt=Decimal("0"),
        max_digits=38,
        decimal_places=18,
    )
    # A blank factor means NOT ESTABLISHED and reaches the engine as None: it calculates
    # with 1, names the factor in a warning and reports REVIEW_REQUIRED (A15 (a)). The
    # schema never fills a default in its place.
    power_factor: ExactDecimal | None = Field(
        default=None,
        gt=Decimal("0"),
        le=Decimal("1"),
        max_digits=38,
        decimal_places=18,
    )
    efficiency: ExactDecimal | None = Field(
        default=None,
        gt=Decimal("0"),
        le=Decimal("1"),
        max_digits=38,
        decimal_places=18,
    )
    utilization_factor: ExactDecimal | None = Field(
        default=None,
        ge=Decimal("0"),
        le=Decimal("1"),
        max_digits=38,
        decimal_places=18,
    )
    demand_factor: ExactDecimal | None = Field(
        default=None,
        ge=Decimal("0"),
        le=Decimal("1"),
        max_digits=38,
        decimal_places=18,
    )
    scenario: LoadScenario = LoadScenario.NORMAL
    power_basis: PowerBasis = PowerBasis.ELECTRICAL_INPUT
    notes: str | None = None

    @model_validator(mode="after")
    def validate_phase_configuration(
        self,
    ) -> Self:
        """Validate phase-specific input requirements."""

        if self.phase_system is PhaseSystem.DC:
            if self.power_factor is not None and self.power_factor != Decimal("1"):
                raise ValueError("DC loads must use a power_factor of 1")
        elif self.power_factor is None:
            raise ValueError("power_factor is required for AC loads")

        return self

    def to_domain(self) -> LoadInput:
        """Convert the validated API request to a domain record."""

        return LoadInput(
            code=self.code,
            name=self.name,
            quantity=self.quantity,
            rated_power_kw=self.rated_power_kw,
            phase_system=self.phase_system,
            voltage_v=self.voltage_v,
            power_factor=self.power_factor,
            efficiency=self.efficiency,
            utilization_factor=self.utilization_factor,
            demand_factor=self.demand_factor,
            scenario=self.scenario,
            power_basis=self.power_basis,
            notes=self.notes,
        )


class LoadGroupCalculationRequest(_RequestBase):
    """API request for calculating an electrical load group."""

    code: str = Field(
        min_length=1,
        max_length=50,
    )
    name: str = Field(
        min_length=1,
        max_length=200,
    )
    loads: tuple[LoadCalculationRequest, ...] = Field(
        min_length=1,
    )
    # Blank means NOT ESTABLISHED, exactly as the per-load factors above.
    coincidence_factor: ExactDecimal | None = Field(
        default=None,
        ge=Decimal("0"),
        le=Decimal("1"),
        max_digits=38,
        decimal_places=18,
    )
    # The load engine has no profile-dependent value; the profile is carried for the
    # result, the run record and the design basis of the project revision.
    jurisdiction_profile: JurisdictionProfile = JurisdictionProfile.IN

    def to_domain(self) -> LoadGroupInput:
        """Convert the validated API request to a domain group."""

        return LoadGroupInput(
            code=self.code,
            name=self.name,
            loads=tuple(load.to_domain() for load in self.loads),
            coincidence_factor=self.coincidence_factor,
        )


class CalculationWarningResponse(_ResponseBase):
    """Structured warning returned by the calculation API."""

    code: LoadWarningCode
    message: str


class LoadCalculationResponse(_ResponseBase):
    """API response for one calculated electrical load."""

    load_code: str
    load_name: str
    scenario: LoadScenario
    phase_system: PhaseSystem
    connected_power_kw: Decimal
    utilized_power_kw: Decimal
    demand_power_kw: Decimal
    apparent_power_kva: Decimal
    reactive_power_kvar: Decimal
    design_current_a: Decimal
    status: CalculationStatus
    warnings: tuple[CalculationWarningResponse, ...]


class LoadGroupCalculationResponse(_ResponseBase):
    """API response for an aggregated load group."""

    group_code: str
    group_name: str
    # The coincidence factor the aggregation used; 1 when it was not established, which
    # the matching COINCIDENCE_FACTOR_NOT_ESTABLISHED warning states.
    coincidence_factor: Decimal
    connected_power_kw: Decimal
    pre_coincidence_demand_kw: Decimal
    demand_power_kw: Decimal
    apparent_power_kva: Decimal
    reactive_power_kvar: Decimal
    load_results: tuple[LoadCalculationResponse, ...]
    status: CalculationStatus
    warnings: tuple[CalculationWarningResponse, ...]
    # Declared engineering assumptions of the aggregation (A15 (e)).
    assumptions: tuple[str, ...]
    jurisdiction_profile: JurisdictionProfile

    @classmethod
    def from_domain(
        cls,
        result: LoadGroupCalculationResult,
        *,
        jurisdiction_profile: JurisdictionProfile,
    ) -> Self:
        """
        Create the response from the domain result and the requested profile.

        The load engine carries no jurisdiction profile, so the profile of the
        request is echoed here rather than read from the result.
        """

        return cls(
            group_code=result.group_code,
            group_name=result.group_name,
            coincidence_factor=result.coincidence_factor,
            connected_power_kw=result.connected_power_kw,
            pre_coincidence_demand_kw=result.pre_coincidence_demand_kw,
            demand_power_kw=result.demand_power_kw,
            apparent_power_kva=result.apparent_power_kva,
            reactive_power_kvar=result.reactive_power_kvar,
            load_results=tuple(
                LoadCalculationResponse.model_validate(load_result)
                for load_result in result.load_results
            ),
            status=result.status,
            warnings=tuple(
                CalculationWarningResponse.model_validate(warning) for warning in result.warnings
            ),
            assumptions=result.assumptions,
            jurisdiction_profile=jurisdiction_profile,
        )


__all__ = [
    "CalculationWarningResponse",
    "LoadCalculationRequest",
    "LoadCalculationResponse",
    "LoadGroupCalculationRequest",
    "LoadGroupCalculationResponse",
]

File: frontend/src/services/loadDemand.ts
import { z } from "zod";

import { calculationRunSummarySchema } from "./calculationRun";
import { jurisdictionProfileSchema } from "./cableContract";
import { exactDecimalSchema } from "./faultContract";
import { ApiError } from "./http";
import {
  AC_POWER_FACTOR_REQUIRED,
  loadCalculationStatusSchema,
  loadScenarioSchema,
  loadWarningCodeSchema,
  phaseSystemSchema,
  powerBasisSchema,
} from "./loadDemandContract";

const optionalRequestText = (maxLength: number) =>
  z.string().trim().max(maxLength).nullable().optional();

// A factor left out means NOT ESTABLISHED, never 1: the engine calculates with 1,
// names the factor in a warning and reports REVIEW_REQUIRED (Master Prompt A15 (a)).
// The form must therefore omit a blank field instead of sending "1".
const optionalFactor = exactDecimalSchema.optional();

export const loadRequestSchema = z
  .object({
    code: z.string().trim().min(1).max(50),
    name: z.string().trim().min(1).max(200),
    quantity: z.number().int().positive(),
    rated_power_kw: exactDecimalSchema,
    phase_system: phaseSystemSchema,
    voltage_v: exactDecimalSchema,
    power_factor: optionalFactor,
    efficiency: optionalFactor,
    utilization_factor: optionalFactor,
    demand_factor: optionalFactor,
    scenario: loadScenarioSchema.optional(),
    power_basis: powerBasisSchema.optional(),
    notes: optionalRequestText(1000),
  })
  .strict()
  .superRefine((values, ctx) => {
    // Mirrors LoadCalculationRequest.validate_phase_configuration: an AC load must
    // state its power factor, because a silent unity factor understates the current.
    if (values.phase_system === "DC") {
      if (values.power_factor !== undefined && Number(values.power_factor) !== 1) {
        ctx.addIssue({
          code: "custom",
          message: "DC loads must use a power_factor of 1",
          path: ["power_factor"],
        });
      }
      return;
    }

    if (values.power_factor === undefined) {
      // Reported as a missing value rather than a custom rule, so the form's
      // describeValidationIssue turns it into "Load 1 - Power factor is
      // required." instead of pasting the raw sentence after the label.
      // AC_POWER_FACTOR_REQUIRED stays the backend's own 422 wording.
      ctx.addIssue({
        code: "invalid_type",
        expected: "string",
        received: "undefined",
        message: AC_POWER_FACTOR_REQUIRED,
        path: ["power_factor"],
      });
    }
  });

export const loadGroupRequestSchema = z
  .object({
    code: z.string().trim().min(1).max(50),
    name: z.string().trim().min(1).max(200),
    loads: z.array(loadRequestSchema).min(1),
    coincidence_factor: optionalFactor,
    jurisdiction_profile: jurisdictionProfileSchema.optional(),
  })
  .strict();

export const loadWarningSchema = z
  .object({
    code: loadWarningCodeSchema,
    message: z.string(),
  })
  .strict();

export const loadResultSchema = z
  .object({
    load_code: z.string(),
    load_name: z.string(),
    scenario: loadScenarioSchema,
    phase_system: phaseSystemSchema,
    connected_power_kw: z.string(),
    utilized_power_kw: z.string(),
    demand_power_kw: z.string(),
    apparent_power_kva: z.string(),
    reactive_power_kvar: z.string(),
    design_current_a: z.string(),
    status: loadCalculationStatusSchema,
    warnings: z.array(loadWarningSchema),
  })
  .strict();

export const loadGroupResponseSchema = z
  .object({
    group_code: z.string(),
    group_name: z.string(),
    // The factor the aggregation used; "1" when it was not established.
    coincidence_factor: z.string(),
    connected_power_kw: z.string(),
    pre_coincidence_demand_kw: z.string(),
    demand_power_kw: z.string(),
    apparent_power_kva: z.string(),
    reactive_power_kvar: z.string(),
    load_results: z.array(loadResultSchema),
    status: loadCalculationStatusSchema,
    warnings: z.array(loadWarningSchema),
    // Declared engineering assumptions of the aggregation (A15 (e)).
    assumptions: z.array(z.string()),
    jurisdiction_profile: jurisdictionProfileSchema,
  })
  .strict();

export const loadRunCreateRequestSchema = z
  .object({
    study: loadGroupRequestSchema,
    project_revision_id: z.string().uuid().optional(),
    notes: optionalRequestText(2000),
  })
  .strict();

export const loadRunResponseSchema = z
  .object({
    run: calculationRunSummarySchema,
    result: loadGroupResponseSchema,
  })
  .strict();

const errorDetailSchema = z.union([
  z.string(),
  z.array(
    z
      .object({
        loc: z.array(z.union([z.string(), z.number()])),
        msg: z.string(),
        type: z.string(),
      })
      .passthrough(),
  ),
]);

const errorResponseSchema = z.object({
  detail: errorDetailSchema,
});

export type LoadRequest = z.infer<typeof loadRequestSchema>;
export type LoadGroupRequest = z.infer<typeof loadGroupRequestSchema>;
export type LoadWarning = z.infer<typeof loadWarningSchema>;
export type LoadResult = z.infer<typeof loadResultSchema>;
export type LoadGroupResponse = z.infer<typeof loadGroupResponseSchema>;
export type LoadRunCreateRequest = z.infer<typeof loadRunCreateRequestSchema>;
export type LoadRunResponse = z.infer<typeof loadRunResponseSchema>;

function formatApiError(data: unknown, status: number): string {
  const parsed = errorResponseSchema.safeParse(data);

  if (!parsed.success) {
    return `Load calculation failed (HTTP ${status}).`;
  }

  if (typeof parsed.data.detail === "string") {
    return parsed.data.detail;
  }

  return parsed.data.detail
    .map((item) => `${item.loc.join(".")}: ${item.msg}`)
    .join("; ");
}

/**
 * Calculate a load schedule and persist it as a run revision.
 *
 * The persisted study is the load group: several loads with their coincidence
 * factor. The returned summary carries the run ID, engine version, profile,
 * reference status and content hash shown in the traceability panel.
 *
 * With a project revision id the run is stored under that revision of the selected
 * project; without one it belongs to no project. An unchanged study returns the
 * stored revision with HTTP 200 instead of a new one (A11); both are valid here.
 */
export async function createLoadRun(
  payload: LoadGroupRequest,
  signal?: AbortSignal,
  projectRevisionId?: string,
): Promise<LoadRunResponse> {
  const validatedPayload = loadGroupRequestSchema.parse(payload);
  const timeoutSignal = AbortSignal.timeout(30_000);
  const requestSignal = signal
    ? AbortSignal.any([signal, timeoutSignal])
    : timeoutSignal;

  const response = await fetch("/api/v1/electrical/load-demand/runs", {
    method: "POST",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(
      projectRevisionId === undefined
        ? { study: validatedPayload }
        : { study: validatedPayload, project_revision_id: projectRevisionId },
    ),
    cache: "no-store",
    signal: requestSignal,
  });

  const data: unknown = await response.json().catch(() => null);

  if (!response.ok) {
    throw new ApiError(formatApiError(data, response.status), response.status);
  }

  const parsed = loadRunResponseSchema.safeParse(data);

  if (!parsed.success) {
    throw new Error("Unexpected response from the KES Electrical OS load runs API.");
  }

  return parsed.data;
}
